docs_src/design/adr/security/0009-Secure-Bootstrapping.md

```diff
@@ -0,0 +1,295 @@
+and block startup
+until the service was "healthy".
+Unfortunately, this feature was removed in 3.x docker-compose.
+(This feature is also unsuppported in swarm mode as well.)
+
+Snaps have an explicit install phase and Kubernetes PODs have optional init containers.
+In other frameworks, initialization is allowed to run to completion
+prior to application components being started in production mode.
+This functionality does not exist in Docker nor docker-compose.
+
+The current lack of an initialization phase is a blocking
+issue for implementing microservice communication security,
+as critical EdgeX core components that are involved with
+microservice communcation (specifically Consul)
+are being brought up in an insecure configuration.
+(Consul's insecure configuration is will be addressed in a
+[separate ADR](https://github.com/edgexfoundry/edgex-docs/issues/296).)
+
+Activities that are best done in the initialization phase include the following:
+
+* Bootstrapping of crytographic secrets needed by the application.
+* Bootstrapping of database users and passwords.
+* Installation of database schema needed for application logic to function.
+* Initialization of authorization frameworks such as configuring RBAC or ACLs.
+* Other one-time initialization activities.
+
+Workarounds when an installation phase is not present include:
+
+* Perform initialization tasks manually,
+  and manually seed secrets into static configuration files.
+* Ship with known hard-coded secrets in static configuration files.
+* Start in an insecure configuration and remain that way.
+* Provision some secrets at runtime.
+
+EdgeX does not have a manual installation flow,
+and uses a combination of the last three approaches.
+
+The objective of this ADR is to define a framework for Docker-based initialization logic in EdgeX.
+This will enable the removal of certain hard-coded secrets in EdgeX
+and enable certain components (such as Consul) to be started in a secure configuration.
+These improvement are necessary pre-requisites to implementing microservice communication security.
+
+### History
+
+In previous releases, container startup sequencing has been
+primarily been driven by Consul service health checks
+backed healthcheck endpoints of particular services
+or by sentinel files placed in the file system
+when certain intialization milestones are reached.
+
+The implementation has been plagued by several issues:
+
+* Sentinel files are not cleaned up if the framework fails or is shut down.
+  Invalid state left over from previous instantiations of the framework
+  causes difficult-to-resolve race conditions.
+  (Implementation of this ADR will try to remove as many as possible,
+  focusing on those that are used to gate startup.
+  Some use of sentinel files may still be required to indicate
+  completion of initialization steps so that they are not re-done
+  if there is no API-based mechanism to determine if such
+  initialization has been completed.)
+
```

Comment:
I think this implies that if the implementation removes a sentinel file, that an API-based mechanism exists, and the implementation will be adjusted to use it?

Reply:
That was my intent.  Though I only found 2 sentinel files that we added and we can remove both.